File: 2024/11/solution.js
const fs = require("fs");
const path = require("path");

const input = fs
  .readFileSync(path.join(__dirname, "input.txt"), "utf-8")
  .split(" ")
  .map(Number);

const memo = new Map();

function blink(n) {
  if (memo.has(n)) {
    return memo.get(n);
  }
  if (n === 0) {
    memo.set(n, [1]);
    return [1];
  }
  const nString = n.toString();
  if (nString.length % 2 === 0) {
    memo.set(n, [
      Number(nString.slice(0, nString.length / 2)),
      Number(nString.slice(nString.length / 2)),
    ]);
    return [
      Number(nString.slice(0, nString.length / 2)),
      Number(nString.slice(nString.length / 2)),
    ];
  }
  memo.set(n, [n * 2024]);
  return [n * 2024];
}

class CustomMap {
  constructor(arr = []) {
    this.map = new Map();
    this.add(arr, 1);
  }
  add(arr, k) {
    for (const n of arr) {
      this.map.set(n, this.map.has(n) ? this.map.get(n) + k : k);
    }
  }
  get count() {
    return this.map.values().reduce((a, b) => a + b, 0);
  }
}

let curr = new CustomMap(input);
for (let i = 0; i < 75; i++) {
  const temp = new CustomMap();
  for (const [n, k] of curr.map) {
    const res = blink(n);
    temp.add(res, k);
  }
  curr = temp;
}

console.log(curr.count);

File: 2024/11/solution.go
package main

import (
	"fmt"
	"os"
	"path/filepath"
	"runtime"
	"strconv"
	"strings"
)

func main() {
	_, filename, _, _ := runtime.Caller(0)
	dirname := filepath.Dir(filename)
	inputFilePath := filepath.Join(dirname, "input.txt")
	data, err := os.ReadFile(inputFilePath)
	if err != nil {
		panic(err)
	}
	puzzleInput := strings.Split(string(data), " ")
	stoneMap := parseInput(&puzzleInput)
	fmt.Println(solve(stoneMap))
}

func parseInput(puzzleInput *[]string) *map[int]int {
	stoneMap := make(map[int]int)
	for _, s := range *puzzleInput {
		if n, err := strconv.Atoi(s); err == nil {
			stoneMap[n]++
		}
	}
	return &stoneMap
}

func blink(stoneMap *map[int]int) {
	newStoneMap := make(map[int]int)
	for n, k := range *stoneMap {
		if n == 0 {
			newStoneMap[1] += k
		} else if s := strconv.Itoa(n); len(s)%2 == 0 {
			if n1, err := strconv.Atoi(s[0 : len(s)/2]); err == nil {
				newStoneMap[n1] += k
			} else {
				panic(err)
			}
			if n2, err := strconv.Atoi(s[len(s)/2:]); err == nil {
				newStoneMap[n2] += k
			} else {
				panic(err)
			}
		} else {
			newStoneMap[n*2024] += k
		}
	}
	*stoneMap = newStoneMap
}

func nStones(stoneMap *map[int]int) int {
	res := 0
	for _, k := range *stoneMap {
		res += k
	}
	return res
}

func solve(stoneMap *map[int]int) (int, int) {
	for i := 0; i < 25; i++ {
		blink(stoneMap)
	}
	part1 := nStones(stoneMap)
	for i := 25; i < 75; i++ {
		blink(stoneMap)
	}
	part2 := nStones(stoneMap)
	return part1, part2
}
